src/ChangeStream/ResourceStack.php

```diff
@@ -0,0 +1,121 @@
+<?php
+
+/*
+ * This file is part of the puli/repository package.
+ *
+ * (c) Bernhard Schussek <bschussek@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Puli\Repository\ChangeStream;
+
+use OutOfBoundsException;
+use Puli\Repository\Api\Resource\PuliResource;
+use Webmozart\Assert\Assert;
+
+/**
+ * Represents a versioned stack of resources for a given path.
+ *
+ * You can access different versions of a resource using a ChangeStream
+ * that will return you resource stacks.
+ *
+ * @since  1.0
+ *
+ * @author Titouan Galopin <galopintitouan@gmail.com>
+ */
+class ResourceStack
+{
+    /**
+     * @var array
+     */
+    private $stack;
+
+    /**
+     * @param array $stack
+     */
+    public function __construct(array $stack)
+    {
+        Assert::allIsInstanceOf($stack, 'Puli\Repository\Api\Resource\PuliResource');
+        Assert::greaterThanEq(count($stack), 1, 'A ResourceStack cannot be empty');
+
+        $this->stack = array_values($stack);
```

Comment:
You can change this to `array_combine(range(1, count($stack)), $stack)` in order to reindex the stack starting at 1.